### PR TITLE
drivers: spi: support for Nuvoton numaker SPI interface

### PR DESCRIPTION
**Pull request by cyliangtw**
This PR is to add Nuvoton numaker series SPI interface.
It's verified on board: NuMaker-PFM-M467.

### CODE REVIEW

File: drivers/spi/spi_numaker.c [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+	}
+	spi_context_unlock_unconditionally(ctx);
+
+	return 0;
+}
+
+static struct spi_driver_api spi_numaker_driver_api = {.transceive = spi_numaker_transceive,
+						       .release = spi_numaker_release};
+
+int spi_numaker_init(const struct device *dev)
```
Comment (fabiobaltieri) [inline drivers/spi/spi_numaker.c:283]: static

File: tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	};
+	fast@0 {
+		compatible = "test-spi-loopback-fast";
+		reg = <0>;
+		spi-max-frequency = <16000000>;
+	};
+	status = "okay";
+	pinctrl-0 = <&spi2_default>;
+	pinctrl-names = "default";
+/*    cs-gpios = <&gpioa 11 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>; */
```
Comment (fabiobaltieri) [inline tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:33]: Drop it?

File: tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+&pinctrl {
+/* CS/CLK/MISO/MOSI --> EVB's NU5: SS/CLK/MISO/MOSI --> PA11/PA10/PA9/PA8 */
```
Comment (fabiobaltieri) [inline tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:4]: Do you really need the comment? The pin names have that information already, this is one more thing to get stale if you end up copypasting it on new boards. I'd suggest dropping it.

Comment (cyliangtw) [reply tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:4]: Could I only keep `/* EVB's NU5: SS/CLK/MISO/MOSI */` ?
else user can't easy find the pad "NU5" to flywire MISO/MOSI for SPI loopback test.

File: tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay (hunk 2 above)
Comment (fabiobaltieri) [reply tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:33]: If it's for the loopback test, why not just enabling it? There's plenty of tests like this that just picks a pin that makes sense so it's ready to use.

File: tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay (hunk 3 above)
Comment (fabiobaltieri) [reply tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:4]: Oh yeah that makes sense.

File: tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay (hunk 2 above)
Comment (cyliangtw) [reply tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:33]: Agree with your opinion to drop this comment.

File: drivers/spi/spi_numaker.c [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+	/* Set Transfer LSB or MSB first */
+	if ((config->operation) & SPI_TRANSFER_LSB) {
+		SPI_SET_LSB_FIRST(dev_cfg->spi);
+	} else {
+		SPI_SET_MSB_FIRST(dev_cfg->spi);
+	}
+
+	/* full/half duplex */
+	if (((config->operation) & BIT(11))) {
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/spi/spi_numaker.c:106]: Replace by meaningful name SPI_HALF_DUPLEX & SPI_CS_ACTIVE_HIGH.

File: drivers/spi/spi_numaker.c [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+	/* Make SPI module be ready to transfer */
+	if (dev_cfg->is_qspi) {
+		QSPI_Open((QSPI_T *)dev_cfg->spi,
+			  (SPI_OP_MODE_GET(config->operation) == SPI_OP_MODE_SLAVE) ? QSPI_SLAVE
+										    : QSPI_MASTER,
+			  (mode == 0)   ? QSPI_MODE_0
+			  : (mode == 1) ? QSPI_MODE_1
+			  : (mode == 2) ? QSPI_MODE_2
+					: QSPI_MODE_3,
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/spi/spi_numaker.c]: Done

File: drivers/spi/spi_numaker.c [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+	return 0;
+}
+
+static int spi_numaker_txrx(const struct device *dev)
+{
+	struct spi_numaker_data *data = dev->data;
+	const struct spi_numaker_config *dev_cfg = dev->config;
+	struct spi_context *ctx = &data->ctx;
+	uint32_t tx_frame = 0U, rx_frame = 0U;
+	uint8_t word_size = 0, spi_dfs = 0;
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/spi/spi_numaker.c]: yes, it's not necessary.

File: drivers/spi/spi_numaker.c (hunk 1 above)
Comment (cyliangtw) [reply drivers/spi/spi_numaker.c]: Done

File: drivers/spi/spi_numaker.c [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+		break;
+	case 24:
+		spi_dfs = 3;
+		break;
+	case 32:
+		spi_dfs = 4;
+		break;
+	default:
+		spi_dfs = 0;
+		ret = -1;
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/spi/spi_numaker.c]: Refined and finished the update at commit  532e5e4aa72c

File: drivers/spi/spi_numaker.c [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+	struct spi_context ctx;
+};
+
+/*
+ * CPOL/CPHA = 0/0 --> SPI_MODE_0
+ * CPOL/CPHA = 0/1 --> SPI_MODE_1
+ * CPOL/CPHA = 1/0 --> SPI_MODE_2
+ * CPOL/CPHA = 1/1 --> SPI_MODE_3
+ */
+const uint32_t smode_tbl[4] = {
```
Comment (fabiobaltieri) [inline drivers/spi/spi_numaker.c:46]: `static` (below as well)

Comment (cyliangtw) [reply drivers/spi/spi_numaker.c]: Update at commit c7392e8f739

**Review by fabiobaltieri — APPROVED** [review]

**Comment by cyliangtw** [comment]
Hi @tbursztyka,
please take a look.

**Comment by cyliangtw** [comment]
Hi @tbursztyka , @ssekar15 ,
If you are available, please look into this PR.

**Review by ssekar15 — APPROVED** [review]
thanks, looks fine for me

**Pull request merged**